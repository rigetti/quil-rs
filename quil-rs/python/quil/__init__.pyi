# This file is automatically generated by pyo3_stub_gen
# ruff: noqa: E501, F401

import builtins
from . import expression
from . import instructions
from . import program
from . import waveforms

class PickleError(QuilError):
    r"""
    Errors when trying to pickle or deepcopy.
    """
    ...

class QuilError(builtins.Exception):
    r"""
    Base exception type for errors raised by this package.
    """
    ...

class ToQuilStringError(QuilError):
    r"""
    Errors which can occur when converting a Quil item to a string.
    """
    ...

class ValueError(QuilError):
    r"""
    Raised when an argument to a quil function has an inappropriate value.
    """
    ...

