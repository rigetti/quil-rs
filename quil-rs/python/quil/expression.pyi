# This file is automatically generated by pyo3_stub_gen
# ruff: noqa: E501, F401

import builtins
import typing
from quil import QuilError
from quil.instructions import MemoryReference
from enum import Enum

class EvaluationError(QuilError):
    r"""
    Errors that may occur while evaluation an ``Expression``.
    """
    ...

class Expression:
    r"""
    The type of Quil expressions.
    
    Quil expressions take advantage of *structural sharing*; if a Quil expression contains the same
    subexpression twice, such as `x + y` in `(x + y) * (x + y)`, the two children of the `*` node
    will be the *same pointer*.  This is implemented through *interning*, also known as
    *hash-consing*; the recursive references to child nodes are done via [`ArcIntern<Expression>`]s.
    Creating an [`ArcIntern`] from an `Expression` will always return the same pointer for the same
    expression.
    
    The structural sharing means that equality, cloning, and hashing on Quil expressions are all
    very cheap, as they do not need to be recursive: equality of [`ArcIntern`]s is a single-pointer
    comparison, cloning of [`ArcIntern`]s is a pointer copy and an atomic increment, and hashing of
    [`ArcIntern`]s hashes a single pointer.  It is also very cheap to key [`HashMap`]s by an
    [`ArcIntern<Expression>`], which can allow for cheap memoization of operations on `Expression`s.
    
    The structural sharing also means that Quil expressions are fundamentally *immutable*; it is
    *impossible* to get an owned or `&mut` reference to the child `Expression`s of any `Expression`,
    as the use of interning means there may be multiple references to that `Expression` at any time.
    
    Note that when comparing Quil expressions, any embedded NaNs are treated as *equal* to other
    NaNs, not unequal, in contravention of the IEEE 754 spec.
    """
    def __add__(self, other:Expression) -> Expression: ...
    def __getnewargs__(self) -> tuple[MemoryReference | FunctionCallExpression | InfixExpression | complex | PrefixExpression | str]: ...
    def __mul__(self, other:Expression) -> Expression: ...
    def __repr__(self) -> builtins.str: ...
    def __sub__(self, other:Expression) -> Expression: ...
    def __truediv__(self, other:Expression) -> Expression: ...
    def evaluate(self, variables:typing.Mapping[builtins.str, builtins.complex], memory_references:typing.Mapping[builtins.str, typing.Sequence[builtins.float]]) -> builtins.complex:
        r"""
        Evaluate an expression, expecting that it may be fully reduced to a single complex number.
        
        If it cannot be reduced to a complex number, this raises an error.
        """
    def into_simplified(self) -> Expression:
        r"""
        Return an expression derived from this one, simplified as much as possible.
        """
    @staticmethod
    def parse(input:builtins.str) -> Expression:
        r"""
        Parse an ``Expression`` from a string.
        
        Raises a ``ParseExpressionError`` error if the string isn't a valid Quil expression.
        """
    def substitute_variables(self, variable_values:typing.Mapping[builtins.str, Expression]) -> Expression:
        r"""
        Substitute an expression in the place of each matching variable.
        """
    def to_real(self) -> builtins.float:
        r"""
        If this is a number with imaginary part "equal to" zero (of _small_ absolute value), return
        that number. Otherwise, error with an evaluation error of a descriptive type.
        """
    class Address(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> MemoryReference: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:MemoryReference) -> Expression.Address: ...
    
    class FunctionCall(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> FunctionCallExpression: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:FunctionCallExpression) -> Expression.FunctionCall: ...
    
    class Infix(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> InfixExpression: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:InfixExpression) -> Expression.Infix: ...
    
    class Number(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> builtins.complex: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:builtins.complex) -> Expression.Number: ...
    
    class Pi(Expression):
        __match_args__ = ()
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls) -> Expression.Pi: ...
    
    class Prefix(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> PrefixExpression: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:PrefixExpression) -> Expression.Prefix: ...
    
    class Variable(Expression):
        __match_args__ = ("_0",)
        @property
        def _0(self) -> builtins.str: ...
        def __getitem__(self, key:builtins.int) -> typing.Any: ...
        def __len__(self) -> builtins.int: ...
        def __new__(cls, _0:builtins.str) -> Expression.Variable: ...
    

class FunctionCallExpression:
    r"""
    The type of function call Quil expressions, e.g. `sin(e)`.
    
    Quil expressions take advantage of *structural sharing*, which is why the `expression` here is
    wrapped in an [`ArcIntern`]; for more details, see the documentation for [`Expression`].
    
    Note that when comparing Quil expressions, any embedded NaNs are treated as *equal* to other
    NaNs, not unequal, in contravention of the IEEE 754 spec.
    """
    @property
    def expression(self) -> Expression: ...
    @property
    def function(self) -> ExpressionFunction: ...
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __getnewargs__(self) -> tuple[ExpressionFunction, Expression]: ...
    def __hash__(self) -> builtins.int: ...
    def __new__(cls, function:ExpressionFunction, expression:Expression) -> FunctionCallExpression: ...
    def __repr__(self) -> builtins.str: ...

class InfixExpression:
    r"""
    The type of infix Quil expressions, e.g. `e1 + e2`.
    
    Quil expressions take advantage of *structural sharing*, which is why the `left` and `right`
    expressions here are wrapped in [`ArcIntern`]s; for more details, see the documentation for
    [`Expression`].
    
    Note that when comparing Quil expressions, any embedded NaNs are treated as *equal* to other
    NaNs, not unequal, in contravention of the IEEE 754 spec.
    """
    @property
    def left(self) -> Expression: ...
    @property
    def operator(self) -> InfixOperator: ...
    @property
    def right(self) -> Expression: ...
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __getnewargs__(self) -> tuple[Expression, InfixOperator, Expression]: ...
    def __hash__(self) -> builtins.int: ...
    def __new__(cls, left:Expression, operator:InfixOperator, right:Expression) -> InfixExpression: ...
    def __repr__(self) -> builtins.str: ...

class ParseExpressionError(QuilError):
    r"""
    Errors that may occur while parsing an ``Expression``.
    """
    ...

class PrefixExpression:
    r"""
    The type of prefix Quil expressions, e.g. `-e`.
    
    Quil expressions take advantage of *structural sharing*, which is why the `expression` here is
    wrapped in an [`ArcIntern`]; for more details, see the documentation for [`Expression`].
    
    Note that when comparing Quil expressions, any embedded NaNs are treated as *equal* to other
    NaNs, not unequal, in contravention of the IEEE 754 spec.
    """
    @property
    def expression(self) -> Expression: ...
    @property
    def operator(self) -> PrefixOperator: ...
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __getnewargs__(self) -> tuple[PrefixOperator, Expression]: ...
    def __hash__(self) -> builtins.int: ...
    def __new__(cls, operator:PrefixOperator, expression:Expression) -> PrefixExpression: ...
    def __repr__(self) -> builtins.str: ...

class ExpressionFunction(Enum):
    r"""
    A function defined within Quil syntax.
    """
    CIS = ...
    COSINE = ...
    EXPONENT = ...
    SINE = ...
    SQUARE_ROOT = ...

    def __getnewargs__(self) -> tuple[builtins.int]: ...
    def __new__(cls, value:builtins.int) -> ExpressionFunction: ...
    def __repr__(self) -> builtins.str: ...

class InfixOperator(Enum):
    CARET = ...
    PLUS = ...
    MINUS = ...
    SLASH = ...
    STAR = ...

    def __getnewargs__(self) -> tuple[builtins.int]: ...
    def __new__(cls, value:builtins.int) -> InfixOperator: ...
    def __repr__(self) -> builtins.str: ...

class PrefixOperator(Enum):
    PLUS = ...
    MINUS = ...

    def __getnewargs__(self) -> tuple[builtins.int]: ...
    def __new__(cls, value:builtins.int) -> PrefixOperator: ...
    def __repr__(self) -> builtins.str: ...

