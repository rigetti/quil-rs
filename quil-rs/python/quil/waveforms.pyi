# This file is automatically generated by pyo3_stub_gen
# ruff: noqa: E501, F401

import builtins

class BoxcarKernel:
    r"""
    A Boxcar waveform.
    """
    @property
    def phase(self) -> builtins.float:
        r"""
        The phase, in cycles.
        """
    @phase.setter
    def phase(self, value: builtins.float) -> None:
        r"""
        The phase, in cycles.
        """
    @property
    def sample_count(self) -> builtins.int:
        r"""
        Sample count, which must be positive.
        """
    @sample_count.setter
    def sample_count(self, value: builtins.int) -> None:
        r"""
        Sample count, which must be positive.
        """
    @property
    def scale(self) -> builtins.float:
        r"""
        Scale applied to the waveform envelope.
        """
    @scale.setter
    def scale(self, value: builtins.float) -> None:
        r"""
        Scale applied to the waveform envelope.
        """
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __new__(cls, phase:builtins.float, scale:builtins.float, sample_count:builtins.int) -> BoxcarKernel:
        r"""
        Create a new `BoxcarKernel`.
        
        This raises an error if `sample_count` is zero.
        """
    def __repr__(self) -> builtins.str: ...
    def into_iq_value(self) -> builtins.complex:
        r"""
        Get the `complex` value this `BoxcarKernel` represents.
        """

class DragGaussian:
    r"""
    Creates a waveform with a DRAG-corrected Gaussian shape.
    
    This is a Gaussian shape with an additional component proportional to the time derivative of the main Gaussian pulse.
    
    See Motzoi F. et al., Phys. Rev. Lett., 103 (2009) 110501. for details.
    """
    @property
    def alpha(self) -> builtins.float:
        r"""
        DRAG parameter - controls strength of the imaginary term
        """
    @alpha.setter
    def alpha(self, value: builtins.float) -> None:
        r"""
        DRAG parameter - controls strength of the imaginary term
        """
    @property
    def anh(self) -> builtins.float:
        r"""
        Qubit anharmonicity - sets rate of evolution for the imaginary term (Hz)
        """
    @anh.setter
    def anh(self, value: builtins.float) -> None:
        r"""
        Qubit anharmonicity - sets rate of evolution for the imaginary term (Hz)
        """
    @property
    def detuning(self) -> builtins.float:
        r"""
        Explicit detuning to bake into iq values
        """
    @detuning.setter
    def detuning(self, value: builtins.float) -> None:
        r"""
        Explicit detuning to bake into iq values
        """
    @property
    def duration(self) -> builtins.float:
        r"""
        Full duration of the pulse (s)
        """
    @duration.setter
    def duration(self, value: builtins.float) -> None:
        r"""
        Full duration of the pulse (s)
        """
    @property
    def fwhm(self) -> builtins.float:
        r"""
        Full width half maximum of the pulse (s)
        """
    @fwhm.setter
    def fwhm(self, value: builtins.float) -> None:
        r"""
        Full width half maximum of the pulse (s)
        """
    @property
    def phase(self) -> builtins.float:
        r"""
        Phase shift for entire waveform
        """
    @phase.setter
    def phase(self, value: builtins.float) -> None:
        r"""
        Phase shift for entire waveform
        """
    @property
    def sample_rate(self) -> builtins.float:
        r"""
        Generate samples at this rate (Hz)
        """
    @sample_rate.setter
    def sample_rate(self, value: builtins.float) -> None:
        r"""
        Generate samples at this rate (Hz)
        """
    @property
    def scale(self) -> builtins.float:
        r"""
        Scale to apply to waveform envelope
        """
    @scale.setter
    def scale(self, value: builtins.float) -> None:
        r"""
        Scale to apply to waveform envelope
        """
    @property
    def t0(self) -> builtins.float:
        r"""
        Center/offset for pulse centroid (s)
        """
    @t0.setter
    def t0(self, value: builtins.float) -> None:
        r"""
        Center/offset for pulse centroid (s)
        """
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __new__(cls, duration:builtins.float, fwhm:builtins.float, t0:builtins.float, anh:builtins.float, alpha:builtins.float, sample_rate:builtins.float, scale:builtins.float, phase:builtins.float, detuning:builtins.float) -> DragGaussian: ...
    def __repr__(self) -> builtins.str: ...
    def into_iq_value(self) -> builtins.list[builtins.complex]:
        r"""
        Get the a list of `complex` values from this waveform.
        """

class ErfSquare:
    r"""
    A waveform with a flat top and edges that are error functions (erfs).
    """
    @property
    def detuning(self) -> builtins.float:
        r"""
        Explicit detuning to bake into iq values
        """
    @detuning.setter
    def detuning(self, value: builtins.float) -> None:
        r"""
        Explicit detuning to bake into iq values
        """
    @property
    def duration(self) -> builtins.float:
        r"""
        Full duration of the pulse (s)
        """
    @duration.setter
    def duration(self, value: builtins.float) -> None:
        r"""
        Full duration of the pulse (s)
        """
    @property
    def pad_left(self) -> builtins.float:
        r"""
        Length of zero padding to add to beginning of pulse (s)
        """
    @pad_left.setter
    def pad_left(self, value: builtins.float) -> None:
        r"""
        Length of zero padding to add to beginning of pulse (s)
        """
    @property
    def pad_right(self) -> builtins.float:
        r"""
        Length of zero padding to add to end of pulse (s)
        """
    @pad_right.setter
    def pad_right(self, value: builtins.float) -> None:
        r"""
        Length of zero padding to add to end of pulse (s)
        """
    @property
    def phase(self) -> builtins.float:
        r"""
        Phase shift for entire waveform
        """
    @phase.setter
    def phase(self, value: builtins.float) -> None:
        r"""
        Phase shift for entire waveform
        """
    @property
    def positive_polarity(self) -> builtins.bool:
        r"""
        Toggle for positive/negative polarity
        """
    @positive_polarity.setter
    def positive_polarity(self, value: builtins.bool) -> None:
        r"""
        Toggle for positive/negative polarity
        """
    @property
    def risetime(self) -> builtins.float:
        r"""
        Slope of erf shoulders (2x FWHM of erf in s)
        """
    @risetime.setter
    def risetime(self, value: builtins.float) -> None:
        r"""
        Slope of erf shoulders (2x FWHM of erf in s)
        """
    @property
    def sample_rate(self) -> builtins.float:
        r"""
        Generate wavform samples at this rate (Hz)
        """
    @sample_rate.setter
    def sample_rate(self, value: builtins.float) -> None:
        r"""
        Generate wavform samples at this rate (Hz)
        """
    @property
    def scale(self) -> builtins.float:
        r"""
        Scale to apply to waveform envelope
        """
    @scale.setter
    def scale(self, value: builtins.float) -> None:
        r"""
        Scale to apply to waveform envelope
        """
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __new__(cls, duration:builtins.float, risetime:builtins.float, sample_rate:builtins.float, pad_left:builtins.float, pad_right:builtins.float, positive_polarity:builtins.bool, scale:builtins.float, phase:builtins.float, detuning:builtins.float) -> ErfSquare: ...
    def __repr__(self) -> builtins.str: ...
    def into_iq_value(self) -> builtins.list[builtins.complex]:
        r"""
        Get the a list of `complex` values from this waveform.
        """

class Gaussian:
    r"""
    Creates a waveform with a Gaussian shape.
    """
    @property
    def detuning(self) -> builtins.float:
        r"""
        Explicit detuning to bake into IQ values
        """
    @detuning.setter
    def detuning(self, value: builtins.float) -> None:
        r"""
        Explicit detuning to bake into IQ values
        """
    @property
    def duration(self) -> builtins.float:
        r"""
        Full duration of the pulse (s)
        """
    @duration.setter
    def duration(self, value: builtins.float) -> None:
        r"""
        Full duration of the pulse (s)
        """
    @property
    def fwhm(self) -> builtins.float:
        r"""
        Full width half maximum of the pulse (s)
        """
    @fwhm.setter
    def fwhm(self, value: builtins.float) -> None:
        r"""
        Full width half maximum of the pulse (s)
        """
    @property
    def phase(self) -> builtins.float:
        r"""
        Phase shift for entire waveform
        """
    @phase.setter
    def phase(self, value: builtins.float) -> None:
        r"""
        Phase shift for entire waveform
        """
    @property
    def sample_rate(self) -> builtins.float:
        r"""
        Generate waveform samples at this rate (Hz)
        """
    @sample_rate.setter
    def sample_rate(self, value: builtins.float) -> None:
        r"""
        Generate waveform samples at this rate (Hz)
        """
    @property
    def scale(self) -> builtins.float:
        r"""
        Scale to apply to waveform envelope
        """
    @scale.setter
    def scale(self, value: builtins.float) -> None:
        r"""
        Scale to apply to waveform envelope
        """
    @property
    def t0(self) -> builtins.float:
        r"""
        Center/offset for pulse centroid (s)
        """
    @t0.setter
    def t0(self, value: builtins.float) -> None:
        r"""
        Center/offset for pulse centroid (s)
        """
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __new__(cls, duration:builtins.float, fwhm:builtins.float, t0:builtins.float, sample_rate:builtins.float, scale:builtins.float, phase:builtins.float, detuning:builtins.float) -> Gaussian: ...
    def __repr__(self) -> builtins.str: ...
    def into_iq_value(self) -> builtins.list[builtins.complex]:
        r"""
        Get the a list of `complex` values from this waveform.
        """

class HermiteGaussian:
    r"""
    Creates a Hermite Gaussian waveform.
    
    This extends the basic DRAG pulse by adding an additional imaginary term to the pulse envelope consisting of a
    Gaussian pulse modified by the second order Hermite polynomial.
    
    Refer to
    "Effects of arbitrary laser or NMR pulse shapes on population inversion and coherence"
    Warren S. Warren. 81, (1984); doi: 10.1063/1.447644
    for details.
    """
    @property
    def alpha(self) -> builtins.float:
        r"""
        DRAG parameter - controls strength of the imaginary term
        """
    @alpha.setter
    def alpha(self, value: builtins.float) -> None:
        r"""
        DRAG parameter - controls strength of the imaginary term
        """
    @property
    def anh(self) -> builtins.float:
        r"""
        Qubit anharmonicity - sets rate of evolution for the imaginary term
        """
    @anh.setter
    def anh(self, value: builtins.float) -> None:
        r"""
        Qubit anharmonicity - sets rate of evolution for the imaginary term
        """
    @property
    def detuning(self) -> builtins.float:
        r"""
        Explicit detuning to bake into iq values
        """
    @detuning.setter
    def detuning(self, value: builtins.float) -> None:
        r"""
        Explicit detuning to bake into iq values
        """
    @property
    def duration(self) -> builtins.float:
        r"""
        Full duration of the pulse
        """
    @duration.setter
    def duration(self, value: builtins.float) -> None:
        r"""
        Full duration of the pulse
        """
    @property
    def fwhm(self) -> builtins.float:
        r"""
        Full width half maximum of the pulse
        """
    @fwhm.setter
    def fwhm(self, value: builtins.float) -> None:
        r"""
        Full width half maximum of the pulse
        """
    @property
    def phase(self) -> builtins.float:
        r"""
        Phase shift for entire waveform
        """
    @phase.setter
    def phase(self, value: builtins.float) -> None:
        r"""
        Phase shift for entire waveform
        """
    @property
    def sample_rate(self) -> builtins.float:
        r"""
        Generate samples at this rate
        """
    @sample_rate.setter
    def sample_rate(self, value: builtins.float) -> None:
        r"""
        Generate samples at this rate
        """
    @property
    def scale(self) -> builtins.float:
        r"""
        Scale to apply to waveform envelope
        """
    @scale.setter
    def scale(self, value: builtins.float) -> None:
        r"""
        Scale to apply to waveform envelope
        """
    @property
    def second_order_hrm_coeff(self) -> builtins.float:
        r"""
        Coefficient of the second order Hermite polynomial term.
        """
    @second_order_hrm_coeff.setter
    def second_order_hrm_coeff(self, value: builtins.float) -> None:
        r"""
        Coefficient of the second order Hermite polynomial term.
        """
    @property
    def t0(self) -> builtins.float:
        r"""
        Center/offset for pulse centroid
        """
    @t0.setter
    def t0(self, value: builtins.float) -> None:
        r"""
        Center/offset for pulse centroid
        """
    def __eq__(self, other:builtins.object) -> builtins.bool: ...
    def __new__(cls, duration:builtins.float, fwhm:builtins.float, t0:builtins.float, anh:builtins.float, alpha:builtins.float, sample_rate:builtins.float, second_order_hrm_coeff:builtins.float, scale:builtins.float, phase:builtins.float, detuning:builtins.float) -> HermiteGaussian: ...
    def __repr__(self) -> builtins.str: ...
    def into_iq_value(self) -> builtins.list[builtins.complex]:
        r"""
        Get the a list of `complex` values from this waveform.
        """

def apply_phase_and_detuning(iq_values:list, phase:builtins.float, detuning:builtins.float, sample_rate:builtins.float) -> None:
    r"""
    Modulate and phase shift waveform IQ data in place.
    """

