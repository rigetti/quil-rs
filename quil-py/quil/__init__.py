"""
The `quil` package provides tools for constructing, manipulating, parsing, and printing [Quil](https://github.com/quil-lang/quil) programs.

⚠️ This package is still in early development and breaking changes should be expected between minor versions.
"""
from .quil import *
