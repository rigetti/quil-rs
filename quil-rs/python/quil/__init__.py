"""The `quil` package provides tools for constructing, manipulating, parsing, and printing [Quil](https://github.com/quil-lang/quil) programs.

⚠️ This package is still in early development and breaking changes should be expected between minor versions.
"""

# The following code exposes the _quil package contents under the namespace quil.
from . import _quil
assert isinstance(_quil.__all__, list) and all(isinstance(s, str) for s in _quil.__all__)
exec(f"from ._quil import {', '.join(_quil.__all__)}; __all__ = {_quil.__all__}")
del _quil
