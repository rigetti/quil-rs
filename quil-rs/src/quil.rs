/// A trait to wrap items which represent some construct within the Quil language.
///
/// If you want to serialize an object to string and fail if it can't be represented as valid Quil, then use
/// `to_quil()`. If you want to serialize an object to string infallibly, and can tolerate invalid Quil, then
/// use `to_quil_or_debug()`.
pub trait Quil: std::fmt::Debug {
    /// Return a string in valid Quil syntax or an error if the item cannot be represented with valid Quil.
    fn to_quil(&self) -> Result<String, ToQuilError> {
        let mut buffer = String::new();
        self.write(&mut buffer, false)?;
        Ok(buffer)
    }

    /// Return a string in valid Quil syntax if possible. Any individual component of this object
    /// which cannot be represented in Quil will be replaced with a `Debug` representation of that
    /// component.
    fn to_quil_or_debug(&self) -> String {
        let mut buffer = String::new();
        let _ = self.write(&mut buffer, true);
        buffer
    }

    /// Write the Quil representation of the item to the given writer. If `fall_back_to_debug`
    /// is `true`, then it must not return an error.
    fn write(
        &self,
        writer: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> Result<(), ToQuilError>;
}

pub type ToQuilResult<T> = Result<T, ToQuilError>;

/// Errors which can occur when converting a Quil item to a string.
#[derive(Debug, thiserror::Error, PartialEq)]
#[non_exhaustive]
pub enum ToQuilError {
    #[error("Failed to write Quil: {0}")]
    FormatError(#[from] std::fmt::Error),
    #[error("Label has not yet been resolved")]
    UnresolvedLabelPlaceholder,
    #[error("Qubit has not yet been resolved")]
    UnresolvedQubitPlaceholder,
}

/// Write an iterator of Quil items to the given writer, joined with the provided `joiner`.
pub(crate) fn write_join_quil<'i, I, T>(
    writer: &mut impl std::fmt::Write,
    fall_back_to_debug: bool,
    values: I,
    joiner: &str,
    prefix: &str,
) -> Result<(), ToQuilError>
where
    I: IntoIterator<Item = &'i T>,
    T: Quil + 'i,
{
    let mut iter = values.into_iter();
    if let Some(first) = iter.next() {
        write!(writer, "{prefix}")?;
        first.write(writer, fall_back_to_debug)?;

        for value in iter {
            write!(writer, "{joiner}{prefix}")?;
            value.write(writer, fall_back_to_debug)?;
        }
    }
    Ok(())
}
