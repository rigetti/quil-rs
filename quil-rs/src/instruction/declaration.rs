use std::str::FromStr;

use nom_locate::LocatedSpan;

use crate::{
    parser::{common::parse_memory_reference, lex, ParseError},
    program::{disallow_leftover, SyntaxError},
    quil::Quil,
};

use super::ArithmeticOperand;

#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
pub enum ScalarType {
    Bit,
    Integer,
    Octet,
    Real,
}

impl Quil for ScalarType {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        _fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        use ScalarType::*;
        write!(
            f,
            "{}",
            match self {
                Bit => "BIT",
                Integer => "INTEGER",
                Octet => "OCTET",
                Real => "REAL",
            }
        )
        .map_err(Into::into)
    }
}

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
pub struct Vector {
    pub data_type: ScalarType,
    pub length: u64,
}

impl Vector {
    pub fn new(data_type: ScalarType, length: u64) -> Self {
        Self { data_type, length }
    }
}

impl Quil for Vector {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        self.data_type.write(f, fall_back_to_debug)?;
        write!(f, "[{}]", self.length).map_err(Into::into)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Sharing {
    pub name: String,
    pub offsets: Vec<Offset>,
}

impl Sharing {
    pub fn new(name: String, offsets: Vec<Offset>) -> Self {
        Self { name, offsets }
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Offset {
    pub offset: u64,
    pub data_type: ScalarType,
}

impl Offset {
    pub fn new(offset: u64, data_type: ScalarType) -> Self {
        Self { offset, data_type }
    }
}

impl Quil for Offset {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "{} ", self.offset)?;
        self.data_type.write(f, fall_back_to_debug)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Declaration {
    pub name: String,
    pub size: Vector,
    pub sharing: Option<Sharing>,
}

impl Declaration {
    pub fn new(name: String, size: Vector, sharing: Option<Sharing>) -> Self {
        Self {
            name,
            size,
            sharing,
        }
    }
}

impl Quil for Declaration {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "DECLARE {} ", self.name)?;
        self.size.write(f, fall_back_to_debug)?;
        if let Some(shared) = &self.sharing {
            write!(f, " SHARING {}", shared.name)?;
            if !shared.offsets.is_empty() {
                write!(f, " OFFSET")?;
                for offset in shared.offsets.iter() {
                    write!(f, " ")?;
                    offset.write(f, fall_back_to_debug)?;
                }
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod test_declaration {
    use super::{Declaration, Offset, ScalarType, Sharing, Vector};
    use crate::quil::Quil;
    use insta::assert_snapshot;
    use rstest::rstest;

    #[rstest]
    #[case(
        "Basic Declaration",
        Declaration{
            name: "ro".to_string(),
            size: Vector{data_type: ScalarType::Bit, length: 1},
            sharing: None,

        }
    )]
    #[case(
        "Shared Declaration",
        Declaration{
            name: "ro".to_string(),
            size: Vector{data_type: ScalarType::Integer, length: 2},
            sharing: Some(Sharing{name: "foo".to_string(), offsets: vec![]})
        }
    )]
    #[case(
        "Shared Declaration with Offsets",
        Declaration{
            name: "ro".to_string(),
            size: Vector{data_type: ScalarType::Real, length: 3},
            sharing: Some(Sharing{
                name: "bar".to_string(),
                offsets: vec![
                    Offset{offset: 4, data_type: ScalarType::Bit},
                    Offset{offset: 5, data_type: ScalarType::Bit}
                ]})
        }
    )]
    fn test_display(#[case] description: &str, #[case] declaration: Declaration) {
        insta::with_settings!({
            snapshot_suffix => description,
        }, {
            assert_snapshot!(declaration.to_quil_or_debug())
        })
    }
}

#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct MemoryReference {
    pub name: String,
    pub index: u64,
}

impl MemoryReference {
    pub fn new(name: String, index: u64) -> Self {
        Self { name, index }
    }
}

impl Quil for MemoryReference {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        _fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "{}[{}]", self.name, self.index).map_err(Into::into)
    }
}

impl std::fmt::Display for MemoryReference {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}[{}]", self.name, self.index)
    }
}

impl FromStr for MemoryReference {
    type Err = SyntaxError<Self>;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let input = LocatedSpan::new(s);
        let tokens = lex(input)?;
        disallow_leftover(
            parse_memory_reference(&tokens).map_err(ParseError::from_nom_internal_err),
        )
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Load {
    pub destination: MemoryReference,
    pub source: String,
    pub offset: MemoryReference,
}

impl Load {
    pub fn new(destination: MemoryReference, source: String, offset: MemoryReference) -> Self {
        Self {
            destination,
            source,
            offset,
        }
    }
}

impl Quil for Load {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "LOAD ")?;
        self.destination.write(f, fall_back_to_debug)?;
        write!(f, " {} ", self.source)?;
        self.offset.write(f, fall_back_to_debug)?;
        Ok(())
    }
}

#[derive(Clone, Debug, PartialEq, Hash)]
pub struct Store {
    pub destination: String,
    pub offset: MemoryReference,
    pub source: ArithmeticOperand,
}

impl Store {
    pub fn new(destination: String, offset: MemoryReference, source: ArithmeticOperand) -> Self {
        Self {
            destination,
            offset,
            source,
        }
    }
}

impl Quil for Store {
    fn write(
        &self,
        f: &mut impl std::fmt::Write,
        fall_back_to_debug: bool,
    ) -> crate::quil::ToQuilResult<()> {
        write!(f, "STORE {} ", self.destination)?;
        self.offset.write(f, fall_back_to_debug)?;
        write!(f, " ")?;
        self.source.write(f, fall_back_to_debug)?;
        Ok(())
    }
}
