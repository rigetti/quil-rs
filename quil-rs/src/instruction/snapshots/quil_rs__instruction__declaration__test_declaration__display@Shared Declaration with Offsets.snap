---
source: quil-rs/src/instruction/declaration.rs
expression: declaration.to_string()
---
DECLARE ro REAL[3] SHARING bar OFFSET 4 BIT 5 BIT
